### PR TITLE
feat: [APPSMTH-29] make step up/down arrows in input Number and Currency widgets optional

### LINKED ISSUE #18041: [Feature]: make step up/down arrows in input Number and Currency widgets optional
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

The arrows to in-/decrease a number or currency value is currently always displayed.
image

User should have the option to disable that overlay.

This also prevents right-aligned icon to show up in Number input widget ie. in above, there should be a **%** according to styling settings.

### Why should this be worked on?

Rates are formatted as #.## %, not with % as prefixed. Also, rates are numbers so using Text is not an ideal workaround, especially for validation. 
Lastly, a number or currency is not always incremental - by default, the overlay should be hidden and enabled by the user for the few cases where increasing / decreasing by one unit is useful.

Front logo [Front conversations](https://app.frontapp.com/open/top_5kudo)

### PR DESCRIPTION
**Pull request by gitstart-appsmith**
## Description
 
This PR makes step up/down arrows optional in

- CurrencyInput 
- Input

NB: This would require updating the Appsmith Documentation

Fixes #18041 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?
- Cypress Tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/widgets/constants.ts [hunk 1]
```diff
@@ -272,3 +272,9 @@ export const dateFormatOptions = [
 export type ThemeProp = {
   theme: Theme;
 };
+
+export enum ButtonPosition {
```
Comment (sbalaji1192) [inline app/client/src/widgets/constants.ts:276]: Let's define this inside the baseInput Widget.

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx [hunk 2]
```diff
@@ -468,6 +477,14 @@ class InputWidget extends BaseInputWidget<InputWidgetProps, WidgetState> {
     );
   };
 
+  updateShowStepArrows() {
```
Comment (sbalaji1192) [inline app/client/src/widgets/InputWidgetV2/widget/index.tsx:480]: we need to write a migration to enable this showStepArrows for existing widgets . Can you have a look at this PR https://github.com/appsmithorg/appsmith/pull/18180/files#diff-d4528352502419c7042ff8f1037b71c3e3895dce1474ea5a73317700f6809953R1132 ?

File: app/client/src/widgets/CurrencyInputWidget/widget/index.tsx [hunk 3]
```diff
@@ -416,6 +427,12 @@ class CurrencyInputWidget extends BaseInputWidget<
       conditionalProps.errorMessage = createMessage(FIELD_REQUIRED_ERROR);
     }
 
+    if (this.props.showStepArrows === false) {
```
Comment (sbalaji1192) [inline app/client/src/widgets/CurrencyInputWidget/widget/index.tsx:430]: lets use the truthy outcome in the if block

File: app/client/src/utils/migrations/InputWidgetMigrations.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import { InputTypes } from "widgets/BaseInputWidget/constants";
+import { WidgetProps } from "widgets/BaseWidget";
+import { DSLWidget } from "widgets/constants";
+
+export const migrateInputWidgetShowStepArrows = (
+  currentDSL: DSLWidget,
+): DSLWidget => {
+  return traverseDSLAndMigrate(currentDSL, (widget: WidgetProps) => {
+    if (
+      widget.showStepArrows === undefined &&
```
Comment (sbalaji1192) [inline app/client/src/utils/migrations/InputWidgetMigrations.ts:11]: let's swap the operand positions of `and` condition

Comment (sbalaji1192) [reply app/client/src/utils/migrations/InputWidgetMigrations.ts:11]: Could you move this function to the `CurrencyInputWidgetMigrations.test.ts` file?

File: app/client/src/widgets/BaseInputWidget/constants.ts [hunk 5]
```diff
@@ -6,3 +6,9 @@ export enum InputTypes {
   PASSWORD = "PASSWORD",
   CURRENCY = "CURRENCY",
 }
+
+export enum ButtonPosition {
```
Comment (sbalaji1192) [inline app/client/src/widgets/BaseInputWidget/constants.ts:10]: Can we please rename this to NumberInputStepButtonPosition. ButtonPosition seems to be misleading.

File: app/client/src/widgets/CurrencyInputWidget/widget/index.tsx [hunk 6]
```diff
@@ -430,6 +431,12 @@ class CurrencyInputWidget extends BaseInputWidget<
       conditionalProps.errorMessage = createMessage(FIELD_REQUIRED_ERROR);
     }
 
+    if (Boolean(this.props.showStepArrows)) {
```
Comment (sbalaji1192) [inline app/client/src/widgets/CurrencyInputWidget/widget/index.tsx:434]: No need for boolean call.

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx [hunk 7]
```diff
@@ -553,6 +553,15 @@ class InputWidget extends BaseInputWidget<InputWidgetProps, WidgetState> {
         );
       }
     }
+
+    if (
+      this.props.inputType === InputTypes.NUMBER &&
+      Boolean(this.props.showStepArrows)
```
Comment (sbalaji1192) [inline app/client/src/widgets/InputWidgetV2/widget/index.tsx:559]: No need for boolean call.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/CurrencyInput/CurrencyInput_ShowStepArrows_spec.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+    cy.get(widgetsPage.inputStepArrows).should("exist"); // step arrows should be visible
+  });
+
+  it("3. Toggle test case to validate that currency input widget, stepArrows should be hidden when toggle value is false", () => {
+    // click on the Js
+    cy.get(widgetsPage.toggleShowStepArrows).click({ force: true });
+
+    // Add showStepArrows action and value as false
+    cy.testJsontext("showsteparrows", false);
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/CurrencyInput/CurrencyInput_ShowStepArrows_spec.js:27]: I guess it should be `cy.testJsontext("showsteparrows", {{false}});`

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/CurrencyInput/CurrencyInput_ShowStepArrows_spec.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    cy.get(widgetsPage.inputStepArrows).should("not.exist"); // step arrows should not be visible
+  });
+
+  it("4. Toggle test case to validate that currency input widget, stepArrows should be visible when toggle value is true", () => {
+    // Add showStepArrows action and value as true
+    cy.testJsontext("showsteparrows", true);
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/CurrencyInput/CurrencyInput_ShowStepArrows_spec.js:34]: I guess it should be `cy.testJsontext("showsteparrows", {{true}});`

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/Input/Inputv2_ShowStepArrows_spec.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    cy.get(widgetsPage.inputStepArrows).should("exist"); // step arrows should be visible
+  });
+
+  it("3. Toggle test case to validate that dataType - NUMBER, stepArrows should be hidden when toggle value is false", () => {
+    // click on the Js
+    cy.get(widgetsPage.toggleShowStepArrows).click({ force: true });
+
+    // Add showStepArrows action and value as false
+    cy.testJsontext("showsteparrows", false);
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/Input/Inputv2_ShowStepArrows_spec.js:29]: here too

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/Input/Inputv2_ShowStepArrows_spec.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    cy.get(widgetsPage.inputStepArrows).should("not.exist"); // step arrows should not be visible
+  });
+
+  it("4. Toggle test case to validate that dataType - NUMBER, stepArrows should be visible when toggle value is true", () => {
+    // Add showStepArrows action and add value as true
+    cy.testJsontext("showsteparrows", true);
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/Input/Inputv2_ShowStepArrows_spec.js:36]: here too

**Comment by gitstart-appsmith** [comment]
Apologies for delay in responding, we wanted to QA the code and avoid regression on our end.

**Review by sbalaji1192 — COMMENTED** [review]
The changes look good! Could you rebase the branch as there are some conflicts

**Review by sbalaji1192 — APPROVED** [review]

**Comment by laveena-en** [comment]
Tested and verified this PR for step up/down arrows being optional for input widget (number) and currency input widget:
- Migration
- JS toggle for the property
- Helper text
- functionality in edit and deployed version

**Comment by dilippitchika** [comment]
Thanks @laveena-en i will take a look at it

**Pull request merged**